internal/scanner/scanner_test.go

```diff
@@ -0,0 +1,308 @@
+			s := scanner.New([]byte(tt.s))
+			_, tok, lit := s.ScanWithRegex()
+			if want, got := tt.tok, tok; want != got {
+				t.Errorf("unexpected token -want/+got\n\t- %d\n\t+ %d", want, got)
+			}
+			if want, got := tt.lit, lit; want != got {
+				t.Errorf("unexpected literal -want/+got\n\t- %s\n\t+ %s", want, got)
+			}
+
+			// Expect an EOF token.
+			if _, tok, _ := s.ScanWithRegex(); tok != token.EOF {
+				t.Errorf("expected eof token, got %d", tok)
+			}
+		})
+	}
+}
+
+func TestScanner_Unread(t *testing.T) {
+	s := scanner.New([]byte(`a /hello/`))
+	_, tok, _ := s.ScanWithRegex()
+	if want, got := token.IDENT, tok; want != got {
+		t.Fatalf("unexpected first token: %d", tok)
+	}
+
+	// First unread should read the same ident again.
+	s.Unread()
+
+	_, tok, _ = s.ScanWithRegex()
+	if want, got := token.IDENT, tok; want != got {
+		t.Fatalf("unexpected token after first unread: %d", tok)
+	}
+
+	// Read the next token using the standard scan.
+	_, tok, _ = s.ScanWithRegex()
+	if want, got := token.REGEX, tok; want != got {
+		t.Fatalf("unexpected token after first unread: %d", tok)
+	}
+
+	// Unread should move back to the beginning and scanning without
+	// regex should give us the division operator.
+	s.Unread()
+	_, tok, _ = s.Scan()
+	if want, got := token.DIV, tok; want != got {
+		t.Fatalf("unexpected token after first unread: %d", tok)
+	}
+
+	// Unread twice and scan again should give us the regex again.
+	s.Unread()
+	s.Unread()
+	_, tok, _ = s.ScanWithRegex()
+	if want, got := token.REGEX, tok; want != got {
+		t.Fatalf("unexpected token after first unread: %d", tok)
+	}
+}
+
+func TestScanner_UnreadEOF(t *testing.T) {
+	// Trailing whitespace should cause unread to not reset to the last token
+	// as the token is considered "complete".
+	s := scanner.New([]byte(`a `))
+	_, tok, _ := s.ScanWithRegex()
+	if want, got := token.IDENT, tok; want != got {
+		t.Fatalf("unexpected first token: %d", tok)
+	}
+
+	// Unread should read the ident again because we didn't scan an EOF token.
+	s.Unread()
+
+	_, tok, _ = s.ScanWithRegex()
+	if want, got := token.IDENT, tok; want != got {
+		t.Fatalf("unexpected token after first unread: %d", tok)
+	}
+
+	// Read the next token to skip the whitespace and hit EOF.
+	_, tok, _ = s.ScanWithRegex()
+	if want, got := token.EOF, tok; want != got {
+		t.Fatalf("unexpected token after first unread: %d", tok)
+	}
+
+	// Unread should not reset back to the beginning of the whitespace
+	// because the whitespace was already consumed.
+	s.Unread()
+	_, tok, _ = s.ScanWithRegex()
+	if want, got := token.EOF, tok; want != got {
+		t.Fatalf("unexpected token after first unread: %d", tok)
+	}
+}
+
+func TestScanner_MultipleTokens(t *testing.T) {
+	for _, tt := range []struct {
+		name string
+		s    string
+		want []token.Token
+	}{
+		{
+			name: "variable declaration",
+			s:    `x = 5`,
+			want: []token.Token{
+				token.IDENT,
+				token.ASSIGN,
+				token.INT,
+			},
+		},
```

Comment:
Can you add one longer test case here? Something that looks like a query? That way we have some coverage over a typical query. May just a `from |> range |> filter |> count |> yield` kind of query.

I would also like to see a test case that has an ILLEGAL token somewhere in the middle. Something as simple as `x = 5 @ y = 1` to show that the scanner can continue scanner after seeing an ILLEGAL token. There is a test case below for multi byte illegal tokens, this case is about being able to continue after seeing an illegal token.

Reply:
I have added these tests.